components/kyma-environment-broker/internal/process/provisioning/staged_manager.go

```diff
@@ -0,0 +1,159 @@
+	})
+}
+
+func (s *stage) Steps() []Step {
+	var result []Step
+	for _, s := range s.steps {
+		result = append(result, s.step)
+	}
+	return result
+}
+
+func NewStagedManager(storage storage.Operations, pub event.Publisher, logger logrus.FieldLogger) *StagedManager {
+	return &StagedManager{
+		log:              logger,
+		operationStorage: storage,
+		publisher:        pub,
+	}
+}
+
+func (m *StagedManager) DefineStages(names []string) {
+	m.stages = make([]*stage, len(names))
+	for i, n := range names {
+		m.stages[i] = &stage{name: n, steps: []stepInfo{}}
+	}
+}
+
+func (m *StagedManager) AddStep(stageName string, weight int, step Step) error {
+	for _, s := range m.stages {
+		if s.name == stageName {
+			s.AddStep(weight, step)
+			return nil
+		}
+	}
+	return fmt.Errorf("Stage %s not defined", stageName)
+}
+
+func (m *StagedManager) Execute(operationID string) (time.Duration, error) {
+	operation, err := m.operationStorage.GetProvisioningOperationByID(operationID)
+	if err != nil {
+		m.log.Errorf("Cannot fetch operation from storage: %s", err)
+		return 3 * time.Second, nil
+	}
+
+	logOperation := m.log.WithFields(logrus.Fields{"operation": operationID, "instanceID": operation.InstanceID, "planID": operation.ProvisioningParameters.PlanID})
+	logOperation.Infof("Start process operation steps for GlobalAcocunt=%s, ", operation.ProvisioningParameters.ErsContext.GlobalAccountID)
+
+	var when time.Duration
+	processedOperation := *operation
+
+	for _, stage := range m.stages {
+		if processedOperation.IsStageFinished(stage.name) {
+			continue
+		}
+
+		for _, step := range stage.Steps() {
+			if operation.IsStepDone(step.Name()) {
+				continue
+			}
+			logStep := logOperation.WithField("step", step.Name()).
+				WithField("stage", stage)
+			logStep.Infof("Start step")
+
+			processedOperation, when, err = m.runStep(step, processedOperation, logStep)
+			if err != nil {
+				logStep.Errorf("Process operation failed: %s", err)
+				return 0, err
+			}
+			if processedOperation.State != domain.InProgress {
+				logStep.Infof("Operation %q got status %s. Process finished.", operation.ID, processedOperation.State)
+				return 0, nil
+			}
+
+			// the step needs a retry
+			if when > 0 {
+				return when, nil
+			}
+
+			// mark step processed
+			operation.FinishStep(step.Name())
+		}
+
+		processedOperation, err = m.saveFinishedStage(processedOperation, stage, logOperation)
+		if err != nil {
+			return time.Second, nil
+		}
+	}
+
+	return 0, nil
+}
+
+func (m *StagedManager) saveFinishedStage(operation internal.ProvisioningOperation, s *stage, log logrus.FieldLogger) (internal.ProvisioningOperation, error) {
+	operation.FinishStage(s.name)
+	op, err := m.operationStorage.UpdateProvisioningOperation(operation)
+	if err != nil {
+		log.Infof("Unable to save operation with finished stage %s: %s", s.name, err.Error())
+		return *op, err
+	}
+	log.Infof("Finished stage %s", s.name)
+	return *op, nil
```

Comment:
maybe it is better to try couple of times as [here](https://github.com/kyma-project/control-plane/blob/main/components/provisioner/internal/operations/executor.go#L161) or use `wait` function instead of repeat all stage steps

Reply:
If you think about retrying failed update operation - the storage is doing that